### PR TITLE
Reindex Behaviour Issues

### PR DESCRIPTION
**Pull request by nathandoef**
Previously, if the `$reindex` operation failed with a `ResourceVersionConflictException` the related  batch job would fail. This has been corrected by adding 10 retry attempts for transactions that have  failed with a `ResourceVersionConflictException` during the `$reindex` operation. 

In addition, the `ResourceIdListStep` was submitting one more resource than expected (i.e. 1001 records processed during a `$reindex` operation if only 1000 `Resources` were in the database). This has been corrected.

### CODE REVIEW

**Review by lukedegruchy — APPROVED** [review]
I worked briefly with the retry interceptor code and I can say this is a necessary and very elegant enhancement/refactoring.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ConcurrentWriteTest.java [hunk 1]
```diff
@@ -824,4 +819,9 @@ public void testTransactionWithCreateClientAssignedIdAndReferenceToThatId() {
 
 	}
 
+	private void setupRetryBehaviour(ServletRequestDetails theServletRequestDetails) {
+		when(theServletRequestDetails.isRetry()).thenReturn(true);
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ConcurrentWriteTest.java:823]: I like how this change cleans up the test

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStep.java [hunk 2]
```diff
@@ -100,12 +100,12 @@ public RunOutcome run(@Nonnull StepExecutionDetails<PT, IT> theStepExecutionDeta
 			previousLastTime = nextChunk.getLastDate().getTime();
 			nextStart = nextChunk.getLastDate();
 
-			while (idBuffer.size() >= MAX_BATCH_OF_IDS) {
+			while (idBuffer.size() > MAX_BATCH_OF_IDS) {
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStep.java:103]: good catch!

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/UserRequestRetryVersionConflictsInterceptor.java [hunk 3]
```diff
@@ -46,39 +47,26 @@
...
 
+	/** Deprecated and moved to {@link ca.uhn.fhir.rest.api.Constants#HEADER_RETRY_ON_VERSION_CONFLICT} */
+	@Deprecated
 	public static final String HEADER_NAME = "X-Retry-On-Version-Conflict";
+
+	/** Deprecated and moved to {@link ca.uhn.fhir.rest.api.Constants#HEADER_MAX_RETRIES} */
+	@Deprecated
 	public static final String MAX_RETRIES = "max-retries";
+
+	/** Deprecated and moved to {@link ca.uhn.fhir.rest.api.Constants#HEADER_RETRY} */
```
Comment (fil512) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/UserRequestRetryVersionConflictsInterceptor.java:58]: excellent.  Use `@deprecated` in the javadoc

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ConcurrentWriteTest.java (hunk 1 above)
Comment (nathandoef) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4ConcurrentWriteTest.java:823]: Yeah moving this into `RequestDetails` was a great idea

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStep.java (hunk 2 above)
Comment (nathandoef) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStep.java:103]: Thanks!

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/UserRequestRetryVersionConflictsInterceptor.java [hunk 4]
```diff
@@ -46,39 +47,26 @@
 @Interceptor
 public class UserRequestRetryVersionConflictsInterceptor {
 
+	/** Deprecated and moved to {@link ca.uhn.fhir.rest.api.Constants#HEADER_RETRY_ON_VERSION_CONFLICT} */
+	@Deprecated
```
Comment (michaelabuckley) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/UserRequestRetryVersionConflictsInterceptor.java:51]: Is Constants visible here?  If so, why not change the definition here?

Comment (nathandoef) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/interceptor/UserRequestRetryVersionConflictsInterceptor.java:51]: Agreed. I was in the middle of bumping versions, but I can submit a new PR for this

**Pull request merged**